components/style/traversal.rs

```diff
@@ -206,61 +314,71 @@ pub fn style_element_in_display_none_subtree<'a, E, C, F>(element: E,
+                                    element: E,
+                                    mut data: &mut AtomicRefMut<ElementData>)
     where E: TElement,
           C: StyleContext<'a>,
           D: DomTraversalContext<E::ConcreteNode>
 {
-    let mode = element.styling_mode();
-    let should_compute = element.borrow_data().map_or(true, |d| d.get_current_styles().is_none());
-    debug!("recalc_style_at: {:?} (should_compute={:?} mode={:?}, data={:?})",
-           element, should_compute, mode, element.borrow_data());
+    debug_assert!(data.as_restyle().map_or(true, |r| r.snapshot.is_none()),
+                  "Snapshots should be expanded by the caller");
 
-    let (computed_display_none, propagated_hint) = if should_compute {
-        compute_style::<_, _, D>(context, data, element)
-    } else {
-        (false, StoredRestyleHint::empty())
+    let compute_self = !data.has_current_styles();
+    let mut inherited_style_changed = false;
+
+    debug!("recalc_style_at: {:?} (compute_self={:?}, dirty_descendants={:?}, data={:?})",
+           element, compute_self, element.has_dirty_descendants(), data);
+
+    // Compute style for this element if necessary.
+    if compute_self {
+        inherited_style_changed = compute_style::<_, _, D>(context, &mut data, traversal_data, element);
+    }
+
+    // Now that matching and cascading is done, clear the bits corresponding to
+    // those operations and compute the propagated restyle hint.
+    let empty_hint = StoredRestyleHint::empty();
+    let propagated_hint = match data.as_restyle_mut() {
+        None => empty_hint,
+        Some(r) => {
+            r.recascade = false;
+            mem::replace(&mut r.hint, empty_hint).propagate()
+        },
     };
+    debug_assert!(data.has_current_styles());
+    trace!("propagated_hint={:?}, inherited_style_changed={:?}", propagated_hint, inherited_style_changed);
 
-    // Preprocess children, computing restyle hints and handling sibling relationships.
-    //
-    // We don't need to do this if we're not traversing children, or if we're performing
-    // initial styling.
-    let will_traverse_children = !computed_display_none &&
-                                 (mode == StylingMode::Restyle ||
-                                  mode == StylingMode::Traverse);
-    if will_traverse_children {
-        preprocess_children::<_, _, D>(context, element, propagated_hint,
-                                       mode == StylingMode::Restyle);
+    // Preprocess children, propagating restyle hints and handling sibling relationships.
+    if !data.styles().is_display_none() &&
+       (element.has_dirty_descendants() || !propagated_hint.is_empty() || inherited_style_changed) {
+        preprocess_children::<_, _, D>(context, element, propagated_hint, inherited_style_changed);
     }
 }
-
+// FIXME(bholley): This should differentiate between matching and cascading,
+// since we have separate bits for each now.
```

Comment:
Add a comment saying that this function returns whether the parent inherited style changes.